packages/@aws-cdk/aws-codestarnotifications/lib/notification-rule.ts

```diff
@@ -0,0 +1,202 @@
+   * @see https://docs.aws.amazon.com/dtconsole/latest/userguide/concepts.html#concepts-api
+   *
+   * @default - No events.
+   */
+  readonly events?: string[];
+
+  /**
+   * The Amazon Resource Name (ARN) of the resource to associate with the notification rule.
+   * Currently, Supported sources include pipelines in AWS CodePipeline and build projects in AWS CodeBuild in this L2 constructor.
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codestarnotifications-notificationrule.html#cfn-codestarnotifications-notificationrule-resource
+   */
+  readonly source: INotificationRuleSource;
+
+  /**
+   * The targets to register for the notification destination.
+   *
+   * @default - No targets are added to the rule. Use `addTarget()` to add a target.
+   */
+  readonly targets?: INotificationRuleTarget[];
+}
+
+/**
+ * Represents a notification rule
+ */
+export interface INotificationRule extends IResource {
+
+  /**
+   * The ARN of the notification rule (i.e. arn:aws:codestar-notifications:::notificationrule/01234abcde)
+   *
+   * @attribute
+   */
+  readonly notificationRuleArn: string;
+
+  /**
+   * Adds target to notification rule
+   *
+   * @param target The SNS topic or AWS Chatbot Slack target
+   * @returns boolean - return true if it had any effect
+   */
+  addTarget(target: INotificationRuleTarget): boolean;
+}
+
+/**
+ * A new notification rule
+ *
+ * @resource AWS::CodeStarNotifications::NotificationRule
+ */
+export class NotificationRule extends Resource implements INotificationRule {
+  /**
+   * Import an existing notification rule provided an ARN
+   * @param scope The parent creating construct
+   * @param id The construct's name
+   * @param notificationRuleArn Notification rule ARN (i.e. arn:aws:codestar-notifications:::notificationrule/01234abcde)
+   */
+  public static fromNotificationRuleArn(scope: constructs.Construct, id: string, notificationRuleArn: string): INotificationRule {
+    class Import extends Resource implements INotificationRule {
+      readonly notificationRuleArn = notificationRuleArn;
+
+      public addTarget(_target: INotificationRuleTarget): boolean {
+        return false;
+      }
+    }
+
+    return new Import(scope, id);
+  }
+
+  /**
+   * @attribute
+   */
+  readonly notificationRuleArn: string;
+
+  /**
+   * The target config of notification rule
+   */
+  private targets: NotificationRuleTargetConfig[] = [];
+
+  /**
+   * The events of notification rule
+   */
+  private events: string[] = [];
+
+  constructor(scope: constructs.Construct, id: string, props: NotificationRuleProps) {
+    super(scope, id);
+
+    const source = props.source.bindAsNotificationRuleSource(this, this);
+
+    if (props.events) {
+      this.addEvents(props.events);
+    }
+
+    props.targets?.forEach((target) => {
+      this.addTarget(target);
+    });
```

Comment:
Like I said already - if you're `bind()`ing a `target` before you fill any properties of the `INotificationRule` that you're passing there, you need to remove `INotificationRule` from the signature of `bindAsNotificationRuleTarget()`. Just leave `scope: Construct` there (the first argument).

You can rectify this by moving this loop after the line you assign `this.notificationRuleArn = resource.ref;` in.

Reply:
Ok, I will do both.